contracts/PreToken.sol

```diff
@@ -0,0 +1,155 @@
+        require(discount > 0, "discount must be > 0");
+        require(agreements[agreementHash].discount == 0, "agreement must not exist yet");
+
+        agreements[agreementHash] = Agreement(owner, 0, discount, valuationCap);
+        agreementOwners[owner] = agreementHash;
+        allAgreements.push(agreementHash);
+
+        emit NewAgreement(owner, agreementHash, discount, valuationCap);
+    }
+
+    function issueTo(bytes32 agreementHash, uint amount) external restrict("PreTokenSigner") {
+        Agreement storage agreement = agreements[agreementHash];
+        require(agreement.discount > 0, "agreement must exist");
+
+        agreement.balance = agreement.balance.add(amount);
+        totalSupply = totalSupply.add(amount);
+
+        emit Transfer(0x0, agreement.owner, amount);
+    }
+
+    /* Restricted function to allow pretoken signers to fix incorrect issuance */
+    function burnFrom(bytes32 agreementHash, uint amount)
+    public restrict("PreTokenSigner") returns (bool) {
+        Agreement storage agreement = agreements[agreementHash];
+        require(agreement.discount > 0, "agreement must exist"); // this is redundant b/c of next requires but be explicit
+        require(amount > 0, "burn amount must be > 0");
+        require(agreement.balance >= amount, "must not burn more than balance"); // .sub would revert anyways but emit reason
+
+        agreement.balance = agreement.balance.sub(amount);
+        totalSupply = totalSupply.sub(amount);
+
+        emit Transfer(agreement.owner, 0x0, amount);
+        return true;
+    }
+
+    function balanceOf(address owner) public view returns (uint) {
+        return agreements[agreementOwners[owner]].balance;
+    }
+
+    /* function to transfer agreement ownership to other wallet by owner
+        it's in ERC20 form so owners can use standard ERC20 wallet just need to pass full balance as value */
+    function transfer(address to, uint amount) public returns (bool) { // solhint-disable-line no-simple-event-func-name
+        _transfer(msg.sender, to, amount);
+        return true;
+    }
+
+    /* Restricted function to allow pretoken signers to fix if pretoken owner lost keys */
+    function transferAgreement(bytes32 agreementHash, address to)
+    public restrict("PreTokenSigner") returns (bool) {
+        _transfer(agreements[agreementHash].owner, to, agreements[agreementHash].balance);
+        return true;
+    }
+
+    /* private function used by transferAgreement & transfer */
+    function _transfer(address from, address to, uint amount) private {
+        Agreement storage agreement = agreements[agreementOwners[from]];
+        require(agreementOwners[from] != 0x0, "from agreement must exists");
+        require(agreementOwners[to] == 0, "to must not have an agreement");
+        require(to != 0x0, "must not transfer to 0x0");
+        require(amount == agreement.balance, "must transfer full balance");
```

Comment:
The internal `_transfer` method should not have an `amount` param. Move this check into the `transfer` method.